packages/run-protocol/src/vaultFactory/vault.js

```diff
@@ -421,126 +423,103 @@ export const makeInnerVault = (
 
-  const transferRun = seat => {
-    const { vaultSeat } = state;
-    const proposal = seat.getProposal();
-    if (proposal.want.RUN) {
-      seat.incrementBy(
-        vaultSeat.decrementBy(harden({ RUN: proposal.want.RUN })),
-      );
-    } else if (proposal.give.RUN) {
-      // We don't allow debt to be negative, so we'll refund overpayments
-      const currentDebt = getCurrentDebt();
-      const acceptedRun = AmountMath.min(proposal.give.RUN, currentDebt);
-      vaultSeat.incrementBy(seat.decrementBy(harden({ RUN: acceptedRun })));
-    }
+  /**
+   * Calculate the fee, the amount to mint and the resulting debt.
+   * The give and the want together reflect a delta, where typpically
+   * one is zero. In that case, the `fee` will also be zero, so the
+   * simple math works
+   *
+   * @param {Amount} currentDebt
+   * @param {Amount} giveAmount
+   * @param {Amount} wantAmount
+   */
+  const loanFee = (currentDebt, giveAmount, wantAmount) => {
+    const fee = ceilMultiplyBy(wantAmount, manager.getLoanFee());
+    const toMint = AmountMath.add(wantAmount, fee);
+    const newDebt = applyDelta(currentDebt, toMint, giveAmount);
+    return { newDebt, toMint, fee };
   };
 
   /**
-   * Calculate the fee, the amount to mint and the resulting debt
+   * Check whether we can proceed with an `adjustBalances`.
    *
-   * @param {ProposalRecord} proposal
-   * @param {{vault: Amount, client: Amount}} debtAfter
+   * @param {Amount} newCollateralPre
+   * @param {Amount} maxDebtPre
+   * @param {Amount} newCollateral
+   * @param {Amount} newDebt
+   * @returns {boolean}
    */
-  const loanFee = (proposal, debtAfter) => {
-    let newDebt;
-    const currentDebt = getCurrentDebt();
-    let toMint = AmountMath.makeEmpty(debtBrand);
-    let fee = AmountMath.makeEmpty(debtBrand);
-    if (proposal.want.RUN) {
-      fee = ceilMultiplyBy(proposal.want.RUN, manager.getLoanFee());
-      toMint = AmountMath.add(proposal.want.RUN, fee);
-      newDebt = AmountMath.add(currentDebt, toMint);
-    } else if (proposal.give.RUN) {
-      newDebt = AmountMath.subtract(currentDebt, debtAfter.vault);
-    } else {
-      newDebt = currentDebt;
+  const checkRestart = (
+    newCollateralPre,
+    maxDebtPre,
+    newCollateral,
+    newDebt,
+  ) => {
+    if (AmountMath.isGTE(newCollateralPre, newCollateral)) {
+      // The collateral did not go up. If the collateral decreased, we pro-rate maxDebt.
+      // We can pro-rate maxDebt because the quote is either linear (price is
+      // unchanging) or super-linear (also called "convex"). Super-linear is from
+      // AMMs: selling less collateral would mean an even smalle price impact, so
+      // this is a conservative choice. `floorMultiply` because the debt ceiling
+      // should constrain more.
+      const runPerCollateral = makeRatioFromAmounts(
+        maxDebtPre,
+        newCollateralPre,
+      );
+      const maxDebtAfter = floorMultiplyBy(newCollateral, runPerCollateral);
```

Comment:
```suggestion
      // this is a conservative choice.
      const runPerCollateral = makeRatioFromAmounts(
        maxDebtPre,
        newCollateralPre,
      );
      // `floorMultiply` because the debt ceiling should be tight.
      const maxDebtAfter = floorMultiplyBy(newCollateral, runPerCollateral);
```